src/backend/columnar/columnar_storage.c

```diff
@@ -0,0 +1,281 @@
+/*-------------------------------------------------------------------------
+ *
+ * columnar_storage.c
+ *
+ * This file contains function definitions for reading and writing to SMgr to
+ * logical offsets (ignoring page headers).
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ * $Id$
+ *
+ *-------------------------------------------------------------------------
+ */
+
+
+#include "postgres.h"
+
+#include "safe_lib.h"
+
+#include "access/heapam.h"
+#include "access/nbtree.h"
+#include "catalog/pg_am.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "storage/lmgr.h"
+#include "storage/smgr.h"
+#include "utils/guc.h"
+#include "utils/memutils.h"
+#include "utils/rel.h"
+#include "utils/relfilenodemap.h"
+
+#include "columnar/columnar.h"
+#include "columnar/columnar_storage.h"
+
+#define COLUMNAR_METAPAGE_BLOCKNO 0
+#define FIRST_LOGICAL_OFFSET COLUMNAR_BYTES_PER_PAGE
+#define LogicalOffsetIsValid(X) ((X) >= FIRST_LOGICAL_OFFSET)
+
+static void ReadFromBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						  char *buf, uint32 len);
+static void WriteToBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						 char *buf, uint32 len, bool clear);
+
+
+/*
+ * ColumnarMetapageInit - initialize a new metapage in an empty relation to
+ * the given contents.
+ */
+void
+ColumnarMetapageInit(Relation rel, ColumnarMetapage *metapage)
+{
+	LockRelationForExtension(rel, ExclusiveLock);
+
+#ifdef USE_ASSERT_CHECKING
+	RelationOpenSmgr(rel);
+	uint64 nblocks = smgrnblocks(rel->rd_smgr, MAIN_FORKNUM);
+	Assert(nblocks == 0);
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+#endif
+
+	/* create new page */
+	Buffer newBuffer = ReadBuffer(rel, P_NEW);
+	ReleaseBuffer(newBuffer);
+
+	WriteToBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				 (char *) metapage, sizeof(ColumnarMetapage), true);
```

Comment:
would be nice to comment why do we need to clear the metapage (unlike the other places calling `WriteToBlock(clear=false)`)

Reply:
Done